openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,272 @@
+        """save the Signatory."""
+        if self.pk is None:
+            temp_image = self.image
+            self.image = None
+            super(Signatory, self).save(*args, **kwargs)
+            self.image = temp_image
+
+        super(Signatory, self).save(*args, **kwargs)
+
+
+class CertificateTemplate(TimeStampedModel):
+    """
+    Certificate Template model to organize content for certificates.
+    """
+
+    name = models.CharField(max_length=255, db_index=True)
+    content = models.TextField(
+        help_text=_(u"HTML Template content data.")
+    )
+    certificate_type = models.CharField(
+        max_length=255, null=True, blank=True,
+        choices=MODE_CHOICES,
+    )
+    organization_id = models.IntegerField(
+        null=True,
+        blank=True,
+        db_index=True,
+        help_text=_(u'Organization of template.'),
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return u"{name}".format(
+            name=self.name
+        )
+
+
+class AbstractCertificate(AbstractCredential):
+    """
+    Abstract Certificate configuration to support multiple type of certificates
+    i.e. Programs, Courses.
+    """
+    signatory = models.ManyToManyField(Signatory)
+    template = models.ManyToManyField(CertificateTemplate, null=True, blank=True)
+    title = models.CharField(max_length=255, null=True, blank=True)
+
+    def __unicode__(self):
+        """Unicode representation."""
+        return u"{site}".format(
+            site=self.site
+        )
+
+
+class CourseCertificate(AbstractCertificate):
+    """
+    Configuration for Course Certificates.
+    """
+    course_id = CourseKeyField(max_length=255)
+    certificate_type = models.CharField(
+        max_length=255,
+        choices=MODE_CHOICES
+    )
+
+    class Meta(object):
+        # Enforce the constraint that mode and certificate should be unique
+        # together.
+        unique_together = (('course_id', 'certificate_type'),)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return u"{course_id}, {certificate_type}".format(
+            course_id=self.course_id,
+            certificate_type=self.certificate_type,
+        )
+
+
+class ProgramCertificate(AbstractCertificate):
+    """
+    Configuration for Program Certificates.
+    """
+    program_id = models.PositiveIntegerField(
+        db_index=True,
+        unique=True,
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return u"{program_id}".format(
+            program_id=self.program_id
+        )
+
+
+class UserCredential(TimeStampedModel):
+    """
+    Credentials issued to a learner.
+    """
+    AWARDED, REVOKED = (
+        "awarded", "revoked",
+    )
+
+    STATUSES_CHOICES = (
+        (AWARDED, _("awarded")),
+        (REVOKED, _("revoked")),
+
+    )
+
+    username = models.CharField(max_length=255, db_index=True)
+    credential = models.ForeignKey(AbstractCredential)
```

Comment:
You need a GenericForeignKey here here not `ForeignKey` here like:
```python
credential = GenericForeignKey('credential_content_type', 'credential_id')
```
https://docs.djangoproject.com/en/1.8/ref/contrib/contenttypes/#generic-relations